tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp

```diff
@@ -1219,30 +1219,57 @@ bool ASTBuildOperation::addConsumer(SwiftASTConsumerRef Consumer) {
   return true;
 }
 
+/// Returns a build operation that `Consumer` can use, in order of the
+/// following:
+///   1. The latest finished build operation that either exactly matches, or
+///      can be used with snapshots
+///   2. If none, the latest in-progress build operation with the same
+///      conditions
+///   3. `nullptr` otherwise
 ASTBuildOperationRef ASTProducer::getBuildOperationForConsumer(
     SwiftASTConsumerRef Consumer,
     IntrusiveRefCntPtr<llvm::vfs::FileSystem> FileSystem,
     SwiftASTManagerRef Mgr) {
+  ASTBuildOperationRef LatestMatch;
   for (auto &BuildOp : llvm::reverse(BuildOperations)) {
     if (BuildOp->isCancelled()) {
       continue;
     }
-    std::vector<ImmutableTextSnapshotRef> Snapshots;
-    Snapshots.reserve(BuildOp->getFileContents().size());
-    for (auto &FileContent : BuildOp->getFileContents()) {
-      if (FileContent.Snapshot) {
-        Snapshots.push_back(FileContent.Snapshot);
-      }
-    }
+
+    // No point checking for a match, we already have one - we're just looking
+    // for a finished operation that can be used with the file contents of
+    // `BuildOp` at this point (which we will prefer over an incomplete
+    // operation, whether that exactly matches or not).
+    if (LatestMatch && !BuildOp->isFinished())
+      continue;
+
     if (BuildOp->matchesSourceState(FileSystem)) {
-      ++Mgr->Impl.Stats->numASTCacheHits;
-      return BuildOp;
-    } else if (Consumer->canUseASTWithSnapshots(Snapshots)) {
-      ++Mgr->Impl.Stats->numASTsUsedWithSnapshots;
-      return BuildOp;
+      if (BuildOp->isFinished()) {
+        ++Mgr->Impl.Stats->numASTCacheHits;
+        return BuildOp;
+      }
+
+      LatestMatch = BuildOp;
+    } else {
+      std::vector<ImmutableTextSnapshotRef> Snapshots;
+      Snapshots.reserve(BuildOp->getFileContents().size());
+      for (auto &FileContent : BuildOp->getFileContents()) {
+        if (FileContent.Snapshot) {
+          Snapshots.push_back(FileContent.Snapshot);
+        }
+      }
+
+      if (Consumer->canUseASTWithSnapshots(Snapshots)) {
+        if (BuildOp->isFinished()) {
+          ++Mgr->Impl.Stats->numASTsUsedWithSnapshots;
+          return BuildOp;
+        }
+
+        LatestMatch = BuildOp;
+      }
     }
   }
-  return nullptr;
+  return LatestMatch;
```

Comment:
I just noticed. I think you need to add the following to keep track of the stats.

```cpp
if (LatestMatch) {
  ++Mgr->Impl.Stats->numASTsUsedWithSnapshots;
}
```

Reply:
Eh. This can actually be wither `numASTCacheHits` or `numASTsUsedWithSnapshots`. I can add a variable to keep track of which to increment, but honestly I don't think it really matters. Really if we're at this point it's not *really* a cache hit anyway, since the AST is still being built.

Reply:
I think we used to count it as a cache hit. Essentially it hits a cache that just hasn’t been populated yet, right?

Reply:
🤷 updated.